### PR TITLE
Migrator provenance headers and throughput improvement

### PR DESCRIPTION
**Pull request by mmatczuk**
This PR adds provenance headers to migrator enabling two-way syncing. To do that it refactors Franz integration code, improving single core performance in a benchmark (left 4.68, right this).

image

Fixes CON-183
Fixes CON-194

### CODE REVIEW

**Comment by mmatczuk** [comment]
I have piggy backed a change to the dashboard to improve diagrams for short benchmarks.

**Comment by mmatczuk** [comment]
Interesting adding adhoc syncpool makes perf worse

```
migrator-1  | time="2025-10-27T13:54:48Z" level=info msg="rolling stats: 80247.5 msg/sec, 80 MB/sec" @service=redpanda-connect bytes/sec=8.02475e+07 label="" msg/sec=80247.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:54:50Z" level=info msg="rolling stats: 100637 msg/sec, 101 MB/sec" @service=redpanda-connect bytes/sec=1.00637e+08 label="" msg/sec=100637 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:54:52Z" level=info msg="rolling stats: 165960 msg/sec, 166 MB/sec" @service=redpanda-connect bytes/sec=1.6596e+08 label="" msg/sec=165960 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:54:54Z" level=info msg="rolling stats: 173904.5 msg/sec, 174 MB/sec" @service=redpanda-connect bytes/sec=1.739045e+08 label="" msg/sec=173904.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:54:56Z" level=info msg="rolling stats: 189106 msg/sec, 189 MB/sec" @service=redpanda-connect bytes/sec=1.89106e+08 label="" msg/sec=189106 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:54:58Z" level=info msg="rolling stats: 188142 msg/sec, 188 MB/sec" @service=redpanda-connect bytes/sec=1.88142e+08 label="" msg/sec=188142 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:00Z" level=info msg="rolling stats: 173151 msg/sec, 173 MB/sec" @service=redpanda-connect bytes/sec=1.73151e+08 label="" msg/sec=173151 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:02Z" level=info msg="rolling stats: 172606.5 msg/sec, 173 MB/sec" @service=redpanda-connect bytes/sec=1.726065e+08 label="" msg/sec=172606.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:04Z" level=info msg="rolling stats: 180370.5 msg/sec, 180 MB/sec" @service=redpanda-connect bytes/sec=1.803705e+08 label="" msg/sec=180370.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:06Z" level=info msg="rolling stats: 174949 msg/sec, 175 MB/sec" @service=redpanda-connect bytes/sec=1.74949e+08 label="" msg/sec=174949 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:08Z" level=info msg="rolling stats: 177059.5 msg/sec, 177 MB/sec" @service=redpanda-connect bytes/sec=1.770595e+08 label="" msg/sec=177059.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:10Z" level=info msg="rolling stats: 187945.5 msg/sec, 188 MB/sec" @service=redpanda-connect bytes/sec=1.879455e+08 label="" msg/sec=187945.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:12Z" level=info msg="rolling stats: 177537.5 msg/sec, 178 MB/sec" @service=redpanda-connect bytes/sec=1.775375e+08 label="" msg/sec=177537.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:14Z" level=info msg="rolling stats: 162139.5 msg/sec, 162 MB/sec" @service=redpanda-connect bytes/sec=1.621395e+08 label="" msg/sec=162139.5 path=root.output.processors.0
migrator-1  | time="2025-10-27T13:55:16Z" level=info msg="rolling stats: 185786 msg/sec, 186 MB/sec" @service=redpanda-connect bytes/sec=1.85786e+08 label="" msg/sec=185786 path=root.output.processors.0
```

File: internal/impl/kafka/franz_headers.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		} else if n := len(h.Value); n == 0 {
+			msg.MetaSetMut(h.Key, "")
+		} else if n == 1 {
+			msg.MetaSetMut(h.Key, rune(h.Value[0]))
+		} else {
+			msg.MetaSetMut(h.Key, string(h.Value))
+		}
+		keys[i] = h.Key
+	}
+	msg.MetaSetMut("kafka_headers", keys)
```
Comment (josephwoodward) [inline internal/impl/kafka/franz_headers.go]: Would it be worth moving `kafka_headers` into the header as a constant, my rational being that it's used a couple of times and it's part of the packages API so better to make it explicit.

File: internal/impl/kafka/franz_headers.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+import (
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+// AddHeaders stores Kafka record headers in message metadata. Each header value
+// is stored under its key. Empty values are stored as nil, single-byte values
+// as rune, multibyte values as string. The ordered list of keys is stored under
```
Comment (josephwoodward) [inline internal/impl/kafka/franz_headers.go]:
```suggestion
// as rune and multibyte values as string. The ordered list of keys is stored under
```

File: internal/impl/kafka/franz_headers.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		keys[i] = h.Key
+	}
+	msg.MetaSetMut("kafka_headers", keys)
+}
+
+// ExtractHeaders reconstructs Kafka record headers from message metadata.
+// Returns nil if no headers are present. This is the inverse of [AddHeaders].
+func ExtractHeaders(msg *service.Message) []kgo.RecordHeader {
+	headers := extractRawHeaders(msg)
+	if headers != nil {
```
Comment (josephwoodward) [inline internal/impl/kafka/franz_headers.go]: Perhaps inline this so it's clear that `headers`'s scope is only for the duration of the `extractRawHeaders` function call. Then `headers` below on line 61 can be declared as part of the `make` call. ie:

```go
if h := extractRawHeaders(msg); h != nil {
	return h
}
```

File: internal/impl/kafka/franz_headers.go (hunk 1 above)
Comment (mmatczuk) [reply internal/impl/kafka/franz_headers.go]: I used to have `kafka_headers_raw` but removed it.

Comment (mmatczuk) [reply internal/impl/kafka/franz_headers.go]: Shall we add it back?

File: internal/impl/kafka/franz_headers_test.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+			headers: nil,
+		},
+		{
+			name: "single header",
+			headers: []kgo.RecordHeader{
+				{Key: "foo", Value: []byte("bar")},
+			},
+		},
+		{
+			name: "multiple unique headers",
```
Comment (josephwoodward) [inline internal/impl/kafka/franz_headers_test.go:42]: suggestion: I assume this function is also covering the `AddHeaders` functionality? If so how about a test to verify behaviour of adding duplicate headers? What are your feelings on renaming it to something like `TestAddThenExtractHeaders` to make it clearer it's testing both add and extract cases?

File: internal/impl/kafka/franz_writer.go [hunk 5]
```diff
@@ -434,38 +464,84 @@ func (w *FranzWriter) WriteBatch(ctx context.Context, b service.MessageBatch) er
...
+}
 
-		wg.Add(len(records))
-		for i, r := range records {
-			details.Client.Produce(ctx, r, promise)
-			dispatch.TriggerSignal(b[i].Context())
-		}
-		wg.Wait()
+func (w *batchWriter) onRecordProduced(r *kgo.Record, err error) {
+	// Note the order of these operations is important, first set the error if
```
Comment (josephwoodward) [inline internal/impl/kafka/franz_writer.go]:
```suggestion
	// Note: the order of these operations is important, first set the error if
```

File: internal/impl/kafka/franz_headers.go (hunk 1 above)
Comment (josephwoodward) [reply internal/impl/kafka/franz_headers.go]: No need, it's not a blocker. If it was there before and you decided to remove it then I'm fine with that too.

**Comment by mmatczuk** [comment]
Updated @josephwoodward PTAL
